students/DanCornutt/lessons/lesson10/assignment/databse.py

```diff
@@ -0,0 +1,206 @@
+"""HW 10:
+Outputs details of timing for all functions in the program.
+Gather this data and write it to a file called timings.csv.
+The file should contain function name, time taken, and number of records
+    processed, and be called timing.csv.
+
+Timing Class from:
+    https://stackabuse.com/python-metaclasses-and-metaprogramming/
+"""
+
+import types
+import time
+import csv
+import logging
+from pymongo import MongoClient
+
+
+def timefunc(fn, *args, **kwargs):
+
+    def fncomposite(*args, **kwargs):
+        timer = Timer()
+        timer.start()
+        rt = fn(*args, **kwargs)
+        timer.stop()
+        # open CSV append and add timer file
+        with open("timing.csv", 'a', newline='') as csv_file:
+            writer = csv.writer(csv_file, delimiter=',')
+            writer.writerow([fn.__name__, timer.elapsed])  # &items processed
+        return rt
+    # return the composite function
+    return fncomposite
+
+
+class Timer:
+    """Timer class to be used to time event
+    """
+
+    def __init__(self, func=time.perf_counter):
+        self.elapsed = 0.0
+        self._func = func
+        self._start = None
+
+    def start(self):
+        if self._start is not None:
+            raise RuntimeError('Already started')
+        self._start = self._func()
+
+    def stop(self):
+        if self._start is None:
+            raise RuntimeError('Not started')
+        end = self._func()
+        self.elapsed += end - self._start
+        self._start = None
+
+    def reset(self):
+        self.elapsed = 0.0
+
+    @property
+    def running(self):
+        return self._start is not None
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, *args):
+        self.stop()
+
+
+class Timed(type):
+    """Timed MetaClass for timing the events of instanced classes
+    """
+    def __new__(cls, name, bases, attr):
+        for name, value in attr.items():
+            if type(value) is types.FunctionType or type(value) is \
+                    types.MethodType:
+                attr[name] = timefunc(value)
+
+        return super(Timed, cls).__new__(cls, name, bases, attr)
+
+
+class MongoDBConnection():
+    """MongoDB Connection to database"""
+
+    def __init__(self, host='127.0.0.1', port=27017):
+        """ be sure to use the ip address not name for local windows"""
+        self.host = host
+        self.port = port
+        self.connection = None
+
+    def __enter__(self):
+        self.connection = MongoClient(self.host, self.port)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.connection.close()
+
+
+class DataLink(metaclass=Timed):
+
+    def __init__(self):
+        mongo = MongoDBConnection()
+        with mongo:
+            self.db = mongo.connection.media
```

Comment:
I like how you are initiating the db here, so available everywhere in the class.